File: 0994-rotting-oranges/0994-rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        q = deque()
        visited = set()
        for x in range(len(grid)):
            for y in range(len(grid[0])):
                if grid[x][y] == 2:
                    q.append(((x, y), 0))
                    visited.add((x, y))
                elif grid[x][y] == 0:
                    visited.add((x, y))
        max_level = 0
        while q:
            curr, level = q.popleft()
            max_level = max(level, max_level)
            visited.add(curr)
            if len(visited) == len(grid) * len(grid[0]):
                break
            for x, y in ((0, 1), (1, 0), (0, -1), (-1, 0)):
                cd = (curr[0] + x, curr[1] + y)
                if (
                    0 <= cd[0] < len(grid)
                    and 0 <= cd[1] < len(grid[0])
                    and cd not in visited
                ):
                    q.append(((cd, level + 1)))
        if len(visited) != len(grid) * len(grid[0]):
            return -1
        return max_level

File: 0994-rotting-oranges/0994-rotting-oranges.dart
import 'dart:collection';

class Solution {
  int orangesRotting(List<List<int>> grid) {
    var q = Queue();
    var visited = <String>{};
    for (var x = 0; x < grid.length; x++) {
      for (var y = 0; y < grid[0].length; y++) {
        if (grid[x][y] == 2) {
          q.add([
            [x, y],
            0
          ]);
          visited.add('$x$y');
        } else if (grid[x][y] == 0) {
          visited.add('$x$y');
        }
      }
    }
    var maxLevel = 0;
    while (q.isNotEmpty) {
      var curr = q.removeFirst();
      var level = curr[1];
      maxLevel = level > maxLevel ? level : maxLevel;
      visited.add('${curr[0][0]}${curr[0][1]}');
      if (visited.length == grid.length * grid[0].length) {
        break;
      }
      for (var move in [
        [0, 1],
        [1, 0],
        [0, -1],
        [-1, 0]
      ]) {
        var cd = [curr[0][0] + move[0], curr[0][1] + move[1]];
        var cdStr = '${cd[0]}${cd[1]}';
        if (cd[0] >= 0 &&
            cd[0] < grid.length &&
            cd[1] >= 0 &&
            cd[1] < grid[0].length &&
            !visited.contains(cdStr)) {
          q.add([cd, level + 1]);
        }
      }
    }
    if (visited.length != grid.length * grid[0].length) {
      return -1;
    }
    return maxLevel;
  }
}